{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc000af5-5975-4817-aaf4-f337296336ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.local/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n"
     ]
    }
   ],
   "source": [
    "import numpy as  np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "\n",
    "# import liana\n",
    "import liana as li\n",
    "from liana.method import singlecellsignalr, connectome, cellphonedb, natmi, logfc, cellchat, geometric_mean\n",
    "\n",
    "import sc_atlas_helpers as ah\n",
    "#from scanpy_helper_submodule import scanpy_helpers as sh\n",
    "\n",
    "# %%\n",
    "from tqdm.auto import tqdm\n",
    "import contextlib\n",
    "import os\n",
    "import statsmodels.stats.multitest\n",
    "import numpy as np\n",
    "from anndata import AnnData\n",
    "import scipy.sparse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "296296fa-49a8-41da-8a1d-5dc8d364efae",
   "metadata": {},
   "source": [
    "# # Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b186868b-c31b-4c80-85ef-d0aa51092b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fdr_correction(df, pvalue_col=\"pvalue\", *, key_added=\"fdr\", inplace=False):\n",
    "    \"\"\"Adjust p-values in a data frame with test results using FDR correction.\"\"\"\n",
    "    if not inplace:\n",
    "        df = df.copy()\n",
    "\n",
    "    df[key_added] = statsmodels.stats.multitest.fdrcorrection(df[pvalue_col].values)[1]\n",
    "\n",
    "    if not inplace:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0129d21e-d3f8-440e-943c-4220823c53ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%\n",
    "\"\"\"Plotting functions for group comparisons\"\"\"\n",
    "\n",
    "import altair as alt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def plot_lm_result_altair(\n",
    "    df,\n",
    "    p_cutoff=0.1,\n",
    "    p_col=\"fdr\",\n",
    "    x=\"variable\",\n",
    "    y=\"group\",\n",
    "    color=\"coef\",\n",
    "    title=\"heatmap\",\n",
    "    cluster=False,\n",
    "    value_max=None,\n",
    "    configure=lambda x: x.configure_mark(opacity=1),\n",
    "    cmap=\"redblue\",\n",
    "    reverse=True,\n",
    "    domain=lambda x: [-x, x],\n",
    "    order=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Plot a results data frame of a comparison as a heatmap\n",
    "    \"\"\"\n",
    "    df_filtered = df.loc[lambda _: _[p_col] < p_cutoff, :]\n",
    "    df_subset = df.loc[\n",
    "        lambda _: _[x].isin(df_filtered[x].unique()) & _[y].isin(df[y].unique())\n",
    "    ]\n",
    "    if not df_subset.shape[0]:\n",
    "        print(\"No values to plot\")\n",
    "        return\n",
    "\n",
    "    if order is None:\n",
    "        order = \"ascending\"\n",
    "        if cluster:\n",
    "            from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "\n",
    "            values_df = df_subset.pivot(index=y, columns=x, values=color)\n",
    "            order = values_df.columns.values[\n",
    "                leaves_list(\n",
    "                    linkage(values_df.values.T, method=\"average\", metric=\"euclidean\")\n",
    "                )\n",
    "            ]\n",
    "\n",
    "    def _get_significance(fdr):\n",
    "        if fdr < 0.001:\n",
    "            return \"< 0.001\"\n",
    "        elif fdr < 0.01:\n",
    "            return \"< 0.01\"\n",
    "        elif fdr < 0.1:\n",
    "            return \"< 0.1\"\n",
    "        else:\n",
    "            return np.nan\n",
    "\n",
    "    df_subset[\"FDR\"] = pd.Categorical([_get_significance(x) for x in df_subset[p_col]])\n",
    "\n",
    "    if value_max is None:\n",
    "        value_max = max(\n",
    "            abs(np.nanmin(df_subset[color])), abs(np.nanmax(df_subset[color]))\n",
    "        )\n",
    "    # just setting the domain in altair will lead to \"black\" fields. Therefore, we constrain the values themselves.\n",
    "    df_subset[color] = np.clip(df_subset[color], *domain(value_max))\n",
    "    return configure(\n",
    "        alt.Chart(df_subset, title=title)\n",
    "        .mark_rect()\n",
    "        .encode(\n",
    "            x=alt.X(x, sort=order),\n",
    "            y=y,\n",
    "            color=alt.Color(\n",
    "                color,\n",
    "                scale=alt.Scale(scheme=cmap, reverse=reverse, domain=domain(value_max)),\n",
    "            ),\n",
    "        )\n",
    "        + alt.Chart(df_subset.loc[lambda x: ~x[\"FDR\"].isnull()])\n",
    "        .mark_point(color=\"white\", filled=True, stroke=\"black\", strokeWidth=0)\n",
    "        .encode(\n",
    "            x=alt.X(x, sort=order),\n",
    "            y=y,\n",
    "            size=alt.Size(\n",
    "                \"FDR:N\",\n",
    "                scale=alt.Scale(\n",
    "                    domain=[\"< 0.001\", \"< 0.01\", \"< 0.1\"],\n",
    "                    range=4 ** np.array([3, 2, 1]),\n",
    "                ),\n",
    "            ),\n",
    "        )\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6449fef-2941-4ce5-84a7-f81d70fa92ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%\n",
    "from typing import Sequence, Union\n",
    "from anndata import AnnData, ImplicitModificationWarning\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from operator import and_\n",
    "from functools import reduce\n",
    "import warnings\n",
    "\n",
    "\n",
    "def pseudobulk(\n",
    "    adata,\n",
    "    *,\n",
    "    groupby: Union[str, Sequence[str]],\n",
    "    aggr_fun=np.sum,\n",
    "    min_obs=10,\n",
    ") -> AnnData:\n",
    "    \"\"\"\n",
    "    Calculate Pseudobulk of groups\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata\n",
    "        annotated data matrix\n",
    "    groupby\n",
    "        One or multiple columns to group by\n",
    "    aggr_fun\n",
    "        Callback function to calculate pseudobulk. Must be a numpy ufunc supporting\n",
    "        the `axis` attribute.\n",
    "    min_obs\n",
    "        Exclude groups with less than `min_obs` observations\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    New anndata object with same vars as input, but reduced number of obs.\n",
    "    \"\"\"\n",
    "    if isinstance(groupby, str):\n",
    "        groupby = [groupby]\n",
    "\n",
    "    combinations = adata.obs.loc[:, groupby].drop_duplicates()\n",
    "\n",
    "    if adata.is_view:\n",
    "        # for whatever reason, the pseudobulk function is terribly slow when operating on a view.\n",
    "        adata = adata.copy()\n",
    "\n",
    "    # precompute masks\n",
    "    masks = {}\n",
    "    for col in groupby:\n",
    "        masks[col] = {}\n",
    "        for val in combinations[col].unique():\n",
    "            masks[col][val] = adata.obs[col] == val\n",
    "\n",
    "    expr_agg = []\n",
    "    obs = []\n",
    "\n",
    "    for comb in combinations.itertuples(index=False):\n",
    "        mask = reduce(and_, (masks[col][val] for col, val in zip(groupby, comb)))\n",
    "        if np.sum(mask) < min_obs:\n",
    "            continue\n",
    "        expr_row = aggr_fun(adata.X[mask, :], axis=0)\n",
    "        obs_row = comb._asdict()\n",
    "        obs_row[\"n_obs\"] = np.sum(mask)\n",
    "        # convert matrix to array if required (happens when aggregating spares matrix)\n",
    "        try:\n",
    "            expr_row = expr_row.A1\n",
    "        except AttributeError:\n",
    "            pass\n",
    "        obs.append(obs_row)\n",
    "        expr_agg.append(expr_row)\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\", ImplicitModificationWarning)\n",
    "        return AnnData(\n",
    "            X=np.vstack(expr_agg),\n",
    "            var=adata.var,\n",
    "            obs=pd.DataFrame.from_records(obs),\n",
    "        )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f7d52cab-c94b-4820-bb2d-27ce3ccf5549",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%\n",
    "\"\"\"Helper functions for cellphonedb analysis\n",
    "\n",
    "Focuses on differential cellphonedb analysis between conditions.\n",
    "\"\"\"\n",
    "from typing import List, Literal\n",
    "import pandas as pd\n",
    "#from .pseudobulk import pseudobulk\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import altair as alt\n",
    "#from .compare_groups.pl import plot_lm_result_altair\n",
    "#from .util import fdr_correction\n",
    "\n",
    "\n",
    "class CpdbAnalysis:\n",
    "    def __init__(\n",
    "        self, cpdb, adata, *, pseudobulk_group_by: List[str], cell_type_column: str\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Class that handles comparative cellphonedb analysis.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        cpdb\n",
    "            pandas data frame with cellphonedb interactions.\n",
    "            Required columns: `source_genesymbols`, `target_genesymbol`.\n",
    "            You can get this from omnipathdb:\n",
    "            https://omnipathdb.org/interactions/?fields=sources,references&genesymbols=1&databases=CellPhoneDB\n",
    "        adata\n",
    "            Anndata object with the target cells. Will use this to derive mean fraction of expressed cells.\n",
    "            Should contain counts in X.\n",
    "        pseudobulk_group_by\n",
    "            See :func:`scanpy_helper.pseudobulk.pseudobulk`. Pseudobulk is used to compute the mean fraction\n",
    "            of expressed cells by patient\n",
    "        cell_type_column\n",
    "            Column in anndata that contains the cell-type annotation.\n",
    "        \"\"\"\n",
    "        self.cpdb = cpdb\n",
    "        self.cell_type_column = cell_type_column\n",
    "        self._find_expressed_genes(adata, pseudobulk_group_by)\n",
    "\n",
    "    def _find_expressed_genes(self, adata, pseudobulk_group_by):\n",
    "        \"\"\"Compute the mean expression and fraction of expressed cells per cell-type.\n",
    "        This is performed on the pseudobulk level, i..e. the mean of means per patient is calculated.\n",
    "        \"\"\"\n",
    "        pb_fracs = pseudobulk(\n",
    "            adata,\n",
    "            groupby=pseudobulk_group_by + [self.cell_type_column],\n",
    "            aggr_fun=lambda x, axis: np.sum(x > 0, axis) / x.shape[axis],  # type: ignore\n",
    "        )\n",
    "        fractions_expressed = pseudobulk(\n",
    "            pb_fracs, groupby=self.cell_type_column, aggr_fun=np.mean\n",
    "        )\n",
    "        fractions_expressed.obs.set_index(self.cell_type_column, inplace=True)\n",
    "\n",
    "        pb = pseudobulk(\n",
    "            adata,\n",
    "            groupby=pseudobulk_group_by + [self.cell_type_column],\n",
    "        )\n",
    "        sc.pp.normalize_total(pb, target_sum=1e6)\n",
    "        sc.pp.log1p(pb)\n",
    "        pb_mean_cell_type = pseudobulk(\n",
    "            pb, groupby=self.cell_type_column, aggr_fun=np.mean\n",
    "        )\n",
    "        pb_mean_cell_type.obs.set_index(self.cell_type_column, inplace=True)\n",
    "\n",
    "        self.expressed_genes = (\n",
    "            fractions_expressed.to_df()\n",
    "            .melt(ignore_index=False, value_name=\"fraction_expressed\")\n",
    "            .reset_index()\n",
    "            .merge(\n",
    "                pb_mean_cell_type.to_df()\n",
    "                .melt(ignore_index=False, value_name=\"expr_mean\")\n",
    "                .reset_index(),\n",
    "                on=[self.cell_type_column, \"variable\"],\n",
    "            )\n",
    "        )\n",
    "\n",
    "    def significant_interactions(\n",
    "        self,\n",
    "        de_res: pd.DataFrame,\n",
    "        *,\n",
    "        pvalue_col=\"pvalue\",\n",
    "        fc_col=\"log2FoldChange\",\n",
    "        gene_symbol_col=\"gene_id\",\n",
    "        max_pvalue=0.1,\n",
    "        min_abs_fc=1,\n",
    "        adjust_fdr=True,\n",
    "        min_frac_expressed=0.1,\n",
    "        de_genes_mode: Literal[\"ligand\", \"receptor\"] = \"ligand\",\n",
    "    ) -> pd.DataFrame:\n",
    "        \"\"\"\n",
    "        Generates a data frame of differentiall cellphonedb interactions.\n",
    "\n",
    "        This function will extract all known ligands (or receptors, respectively) from a list of differentially expressed\n",
    "        and find all receptors (or ligands, respectively) that are expressed above a certain cutoff in all cell-types.\n",
    "\n",
    "        Parameters:\n",
    "        -----------\n",
    "        de_res\n",
    "            List of differentially expressed genes\n",
    "        pvalue_col\n",
    "            column in de_res that contains the pvalue or false discovery rate\n",
    "        gene_id_col\n",
    "            column in de_res that contains the gene symbol\n",
    "        min_frac_expressed\n",
    "            Minimum fraction cells that need to express the receptor (or ligand) to be considered a potential interaction\n",
    "        de_genes_mode\n",
    "            If the list of de genes provided are ligands (default) or receptors. In case of `ligand`, cell-types\n",
    "            that express corresonding receptors above the threshold will be identified. In case of `receptor`,\n",
    "            cell-types that express corresponding ligands above the threshold will be identified.\n",
    "        adjust_fdr\n",
    "            If True, calculate false discovery rate on the pvalue, after filtering for genes that are contained\n",
    "            in the cellphonedb.\n",
    "        \"\"\"\n",
    "        if de_genes_mode == \"ligand\":\n",
    "            cpdb_de_col = \"source_genesymbol\"\n",
    "            cpdb_expr_col = \"target_genesymbol\"\n",
    "        elif de_genes_mode == \"receptor\":\n",
    "            cpdb_de_col = \"target_genesymbol\"\n",
    "            cpdb_expr_col = \"source_genesymbol\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalud value for de_genes_mode!\")\n",
    "\n",
    "        de_res = de_res.loc[lambda x: x[gene_symbol_col].isin(self.cpdb[cpdb_de_col])]\n",
    "        if adjust_fdr:\n",
    "            de_res = fdr_correction(de_res, pvalue_col=pvalue_col, key_added=\"fdr\")\n",
    "            pvalue_col = \"fdr\"\n",
    "\n",
    "        significant_genes = de_res.loc[\n",
    "            lambda x: (x[pvalue_col] < max_pvalue) & (np.abs(x[fc_col]) >= min_abs_fc),\n",
    "            gene_symbol_col,\n",
    "        ].unique()  # type: ignore\n",
    "        significant_interactions = self.cpdb.loc[\n",
    "            lambda x: x[cpdb_de_col].isin(significant_genes)\n",
    "        ]\n",
    "\n",
    "        res_df = (\n",
    "            self.expressed_genes.loc[\n",
    "                lambda x: x[\"fraction_expressed\"] >= min_frac_expressed\n",
    "            ]  # type: ignore\n",
    "            .merge(\n",
    "                significant_interactions,\n",
    "                left_on=\"variable\",\n",
    "                right_on=cpdb_expr_col,\n",
    "            )\n",
    "            .drop(columns=[\"variable\"])\n",
    "            .merge(de_res, left_on=cpdb_de_col, right_on=gene_symbol_col)\n",
    "            .drop(columns=[gene_symbol_col])\n",
    "        )\n",
    "\n",
    "        return res_df\n",
    "\n",
    "    def plot_result(\n",
    "        self,\n",
    "        cpdb_res,\n",
    "        *,\n",
    "        pvalue_col=\"fdr\",\n",
    "        group_col=\"group\",\n",
    "        fc_col=\"log2FoldChange\",\n",
    "        title=\"CPDB analysis\",\n",
    "        aggregate=True,\n",
    "        clip_fc_at=(-5, 5),\n",
    "        label_limit=100,\n",
    "        cluster: Literal[\"heatmap\", \"dotplot\"] = \"dotplot\",\n",
    "        de_genes_mode: Literal[\"ligand\", \"receptor\"] = \"ligand\",\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Plot cpdb results as heatmap\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        cpdb_res\n",
    "            result of `significant_interactions`. May be further filtered or modified.\n",
    "        group_col\n",
    "            column to be used for the y axis of the heatmap\n",
    "        aggregate\n",
    "            whether to merge multiple targets of the same ligand into a single column\n",
    "        de_genes_mode\n",
    "            If the list of de genes provided are ligands (default) or receptors. If receptor, will show the dotplot\n",
    "            at the top (source are expressed ligands) and the de heatmap at the bottom (target are the DE receptors).\n",
    "            Otherwise the other way round.\n",
    "        \"\"\"\n",
    "        if de_genes_mode == \"ligand\":\n",
    "            cpdb_de_col = \"source_genesymbol\"\n",
    "            cpdb_expr_col = \"target_genesymbol\"\n",
    "        elif de_genes_mode == \"receptor\":\n",
    "            cpdb_de_col = \"target_genesymbol\"\n",
    "            cpdb_expr_col = \"source_genesymbol\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalud value for de_genes_mode!\")\n",
    "\n",
    "        cpdb_res[fc_col] = np.clip(cpdb_res[fc_col], *clip_fc_at)\n",
    "\n",
    "        # aggregate if there are multiple receptors per ligand\n",
    "        if aggregate:\n",
    "            cpdb_res = (\n",
    "                cpdb_res.groupby(\n",
    "                    [\n",
    "                        self.cell_type_column,\n",
    "                        cpdb_de_col,\n",
    "                        fc_col,\n",
    "                        pvalue_col,\n",
    "                        group_col,\n",
    "                    ]\n",
    "                )\n",
    "                .agg(\n",
    "                    n=(cpdb_expr_col, len),\n",
    "                    fraction_expressed=(\"fraction_expressed\", np.max),\n",
    "                    expr_mean=(\"expr_mean\", np.max),\n",
    "                )\n",
    "                .reset_index()\n",
    "                .merge(\n",
    "                    cpdb_res.groupby(cpdb_de_col).agg(\n",
    "                        **{\n",
    "                            cpdb_expr_col: (\n",
    "                                cpdb_expr_col,\n",
    "                                lambda x: \"|\".join(np.unique(x)),\n",
    "                            )\n",
    "                        }\n",
    "                    ),\n",
    "                    on=cpdb_de_col,\n",
    "                )\n",
    "            )\n",
    "\n",
    "        cpdb_res[\"interaction\"] = [\n",
    "            f\"{s}_{t}\" for s, t in zip(cpdb_res[cpdb_de_col], cpdb_res[cpdb_expr_col])\n",
    "        ]\n",
    "\n",
    "        # cluster heatmap\n",
    "        if cluster is not None:\n",
    "            from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "\n",
    "            _idx = self.cell_type_column if cluster == \"dotplot\" else group_col\n",
    "            _values = \"fraction_expressed\" if cluster == \"dotplot\" else fc_col\n",
    "            _columns = \"interaction\"\n",
    "            values_df = (\n",
    "                cpdb_res.loc[:, [_idx, _values, _columns]]\n",
    "                .drop_duplicates()\n",
    "                .pivot(\n",
    "                    index=_idx,\n",
    "                    columns=_columns,\n",
    "                    values=_values,\n",
    "                )\n",
    "                .fillna(0)\n",
    "            )\n",
    "            order = values_df.columns.values[\n",
    "                leaves_list(\n",
    "                    linkage(values_df.values.T, method=\"average\", metric=\"euclidean\")\n",
    "                )\n",
    "            ]\n",
    "        else:\n",
    "            order = \"ascending\"\n",
    "\n",
    "        p1 = plot_lm_result_altair(\n",
    "            cpdb_res,\n",
    "            color=fc_col,\n",
    "            p_col=pvalue_col,\n",
    "            x=\"interaction\",\n",
    "            configure=lambda x: x,\n",
    "            title=\"\",\n",
    "            order=order,\n",
    "            p_cutoff=1,\n",
    "        ).encode(\n",
    "            x=alt.X(\n",
    "                title=None,\n",
    "                axis=alt.Axis(\n",
    "                    labelExpr=\"split(datum.label, '_')[0]\",\n",
    "                    orient=\"top\" if de_genes_mode == \"receptor\" else \"bottom\",\n",
    "                ),\n",
    "            )\n",
    "        )\n",
    "\n",
    "        p2 = (\n",
    "            alt.Chart(cpdb_res)\n",
    "            .mark_circle()\n",
    "            .encode(\n",
    "                x=alt.X(\n",
    "                    \"interaction\",\n",
    "                    axis=alt.Axis(\n",
    "                        grid=True,\n",
    "                        orient=\"bottom\" if de_genes_mode == \"receptor\" else \"top\",\n",
    "                        title=None,\n",
    "                        labelExpr=\"split(datum.label, '_')[1]\",\n",
    "                        labelLimit=label_limit,\n",
    "                    ),\n",
    "                    sort=order,\n",
    "                ),\n",
    "                y=alt.Y(self.cell_type_column, axis=alt.Axis(grid=True), title=None),\n",
    "                size=alt.Size(\"fraction_expressed\"),\n",
    "                color=alt.Color(\"expr_mean\", scale=alt.Scale(scheme=\"cividis\")),\n",
    "            )\n",
    "        )\n",
    "\n",
    "        if de_genes_mode == \"receptor\":\n",
    "            p1, p2 = p2, p1\n",
    "\n",
    "        return (\n",
    "            alt.vconcat(p1, p2, title=title)\n",
    "            .resolve_scale(size=\"independent\", color=\"independent\", x=\"independent\")\n",
    "            .configure_mark(opacity=1)\n",
    "            .configure_concat(spacing=label_limit - 130)\n",
    "        )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8d22b42c-611c-42d6-b209-3c03a5f5f5cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"/data/projects/2022/CRCA/results/v0.1/crc-atlas-dataset/latest/ds_analyses/liana_cell2cell/core_atlas/adata_rank_agregate.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "629b30ad-04b7-4084-89b4-8841958e4b9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = sc.read_h5ad(\"/data/projects/2022/CRCA/results/v0.1/crc-atlas-dataset/latest/ds_analyses/liana_cell2cell/core_atlas/tumor_normal/01_subset_adata/adata_tumor_normal_matched.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "77af6f10-4b4f-4813-82fd-ff65f826260e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Cancer cell',\n",
       " 'DC mature',\n",
       " 'Endothelial',\n",
       " 'Enteroendocrine',\n",
       " 'Epithelial',\n",
       " 'Fibroblast',\n",
       " 'Macrophage',\n",
       " 'Mast cell',\n",
       " 'Monocyte',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Pericyte',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(d.obs.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1f4c7b64-0f51-410c-800e-62600a80d968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Cancer cell',\n",
       " 'Endothelial',\n",
       " 'Fibroblast',\n",
       " 'Macrophage',\n",
       " 'Mast cell',\n",
       " 'Monocyte',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Pericyte',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9ef998ac-1606-406b-b4c8-4f7360ec841d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Cancer cell',\n",
       " 'Endothelial',\n",
       " 'Fibroblast',\n",
       " 'Macrophage',\n",
       " 'Mast cell',\n",
       " 'Monocyte',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Pericyte',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cpdb_sig_int.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "94b7aa77-76fe-4116-a6ad-9d551fe908d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "modified_list = [item.replace(\" \", \"_\") for item in list(set(adata.obs.cell_type))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81e277fe-8bfa-4eb4-94f1-2f677414aadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder containing the files\n",
    "folder_path = \"/data/projects/2022/CRCA/results/v0.1/crc-atlas-dataset/latest/ds_analyses/liana_cell2cell/core_atlas/tumor_normal/03_deseq2/\"\n",
    "\n",
    "# Initialize an empty list to hold all DataFrames\n",
    "df_list = []\n",
    "\n",
    "# Loop through each element in the modified_list\n",
    "for element in modified_list:\n",
    "    # Find the file that starts with the element name\n",
    "    for file_name in os.listdir(folder_path):\n",
    "        if file_name.startswith(element):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            file_path = os.path.join(folder_path, file_name)\n",
    "            df = pd.read_csv(file_path,sep=\"\\t\")\n",
    "\n",
    "            # Add a new column \"cell_name\" with the value of the element\n",
    "            df[\"cell_name\"] = element\n",
    "\n",
    "            # Append the DataFrame to the list\n",
    "            df_list.append(df)\n",
    "            break  # Break after finding the matching file\n",
    "\n",
    "# Concatenate all DataFrames into one\n",
    "final_df = pd.concat(df_list, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7500a1d9-0ad4-4b22-8b0c-ec28156f12f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = \"/data/projects/2022/CRCA/results/v0.1/crc-atlas-dataset/latest/ds_analyses/liana_cell2cell/core_atlas/tumor_normal/03_deseq2/\"\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "de_res = {}\n",
    "for f in Path(de_res_dir).glob(\"*DESeq2_result.tsv\"):\n",
    "    ct = f.name.replace(\"-DESeq2_result.tsv\", \"\").replace(\"_\", \" \")\n",
    "    de_res[ct] = pd.read_csv(f, sep=\"\\t\")\n",
    "\n",
    "# %%\n",
    "de_res_all2 = pd.concat([df.assign(cell_type=ct) for ct, df in de_res.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d895ca78-f426-436b-8ed9-99f2946fb77a",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res_dir=\"/data/projects/2022/CRCA/results/v1/final/liana_cell2cell/core_atlas/tumor_normal/epithelial_cancer/03_deseq2/tumor_paired/deseq2_dgea\"\n",
    "from pathlib import Path\n",
    "\n",
    "de_res = {}\n",
    "for f in Path(de_res_dir).glob(\"*DESeq2_result.tsv\"):\n",
    "    ct = f.name.replace(\"-DESeq2_result.tsv\", \"\").replace(\"_\", \" \")\n",
    "    de_res[ct] = pd.read_csv(f, sep=\"\\t\")\n",
    "\n",
    "# %%\n",
    "de_res_all = pd.concat([df.assign(cell_type=ct) for ct, df in de_res.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "831cc70e-707e-471e-bd8f-1c0e27ca4a2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell memory',\n",
       " 'CD4 cycling',\n",
       " 'CD8 cycling',\n",
       " 'DC mature',\n",
       " 'DC3',\n",
       " 'Endothelial lymphatic',\n",
       " 'Enteroendocrine',\n",
       " 'Epithelial',\n",
       " 'Fibroblast S1',\n",
       " 'ILC',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Plasma IgG',\n",
       " 'Plasmablast',\n",
       " 'Schwann cell',\n",
       " 'Treg',\n",
       " 'cDC1',\n",
       " 'gamma',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(de_res_all.cell_type)\n",
    "\n",
    "'B cell memory','B cell'\n",
    "'Endothelial lymphatic','Endothelial'\n",
    " 'Fibroblast S1','Fibroblast'\n",
    "'Plasma IgG', 'Plasma cell'\n",
    " 'Plasmablast' 'Plasma cell'\n",
    " 'Treg','T cell regulatory'\n",
    " 'DC mature', 'cDC'\n",
    " 'DC3','cDC'\n",
    " 'cDC1','cDC'\n",
    " 'CD4 cycling', 'T cell CD4'\n",
    " 'CD8 cycling','T cell CD8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "097ab217-b33c-4207-b157-52d71204dfc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary to map old cell_type values to new ones\n",
    "rename_dict = {\n",
    "    'B cell memory': 'B cell',\n",
    "    'Endothelial lymphatic': 'Endothelial',\n",
    "    'Fibroblast S1': 'Fibroblast',\n",
    "    'Plasma IgG': 'Plasma cell',\n",
    "    'Plasmablast': 'Plasma cell',\n",
    "    'Treg': 'T cell regulatory',\n",
    "    'DC mature': 'cDC',\n",
    "    'DC3': 'cDC',\n",
    "    'cDC1': 'cDC',\n",
    "    'CD4 cycling': 'T cell CD4',\n",
    "    'CD8 cycling': 'T cell CD8',\n",
    "    'Enteroendocrine':'Endothelial',\n",
    "    'ILC':'T cell CD8',\n",
    "    'gamma':'T cell CD8'\n",
    "}\n",
    "\n",
    "# Replace the values in the \"cell_type\" column based on the dictionary\n",
    "de_res_all['cell_type'] = de_res_all['cell_type'].replace(rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8e6deb5f-0fab-48ed-9331-57aefb7d2d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res_all = de_res_all.rename(columns={'symbol':\"gene_id\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f776fea1-097b-4a95-b4e1-3036b4e3f02c",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res_all2 = de_res_all2.rename(columns={'symbol':\"gene_id\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "652f580f-fdab-408e-9542-f6b0f21b71a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Endothelial',\n",
       " 'Epithelial',\n",
       " 'Fibroblast',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(de_res_all.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "158820f8-c3ab-4243-8ab4-8c412fa88c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Cancer cell',\n",
       " 'Endothelial',\n",
       " 'Fibroblast',\n",
       " 'Macrophage',\n",
       " 'Mast cell',\n",
       " 'Monocyte',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Pericyte',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cpdb_sig_int.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86dd766c-eb15-4927-a410-f2711aebc379",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res = final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c3f9d844-5580-4b6b-a7bf-012ec9699875",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res['cell_name'] = de_res['cell_name'].str.replace('_', ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "72523d96-a65d-442d-9e49-98a4ab51ea6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Endothelial',\n",
       " 'Epithelial',\n",
       " 'Fibroblast',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(de_res_all.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c7f9fe60-15fe-45c0-a200-c8aa11f6af48",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'Cancer cell',\n",
       " 'Endothelial',\n",
       " 'Fibroblast',\n",
       " 'Macrophage',\n",
       " 'Mast cell',\n",
       " 'Monocyte',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Pericyte',\n",
       " 'Plasma cell',\n",
       " 'Schwann cell',\n",
       " 'T cell CD4',\n",
       " 'T cell CD8',\n",
       " 'T cell regulatory',\n",
       " 'cDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f0fcb009-e654-4505-874c-cbe191b16fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# %%\n",
    "#result of `significant_interactions`. May be further filtered or modified.\n",
    "cpdb_res = adata.uns['rank_aggregate'].loc[\n",
    "        lambda x: x[\"specificity_rank\"] <= 0.01\n",
    "    ]\n",
    "\n",
    "# %%\n",
    "# rename columns in liana results \n",
    "cpdb_res=cpdb_res.rename(columns={\"ligand_complex\":\"source_genesymbol\",\"receptor_complex\":\"target_genesymbol\"})\n",
    "\n",
    "# %%\n",
    "cpdb_res.columns\n",
    "\n",
    "# %%\n",
    "# use scanpy helper class CpdbAnalysis to compute pseudobulk, cell fraction and \n",
    "cpdba = CpdbAnalysis(\n",
    "    cpdb_res,\n",
    "    adata,\n",
    "    pseudobulk_group_by=[\"patient_id\"],\n",
    "    cell_type_column=\"cell_type\"\n",
    ")\n",
    "\n",
    "# %%\n",
    "cpdba\n",
    "\n",
    "# %%\n",
    "cpdb_sig_int = cpdba.significant_interactions(\n",
    "    de_res_all2, max_pvalue=0.1\n",
    ")\n",
    "\n",
    "# %%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f99658cc-44ae-489b-9fbb-7a9d2f91e0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#columns_to_replace = ['cell_type', 'source', 'target']\n",
    "#cpdb_sig_int[columns_to_replace] = cpdb_sig_int[columns_to_replace].replace({'cDC': 'Dendritic cell', 'pDC': 'Dendritic cell'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8de170dd-5acb-4df2-aaba-723bd75c49cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type_x</th>\n",
       "      <th>fraction_expressed</th>\n",
       "      <th>expr_mean</th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>source_genesymbol</th>\n",
       "      <th>target_genesymbol</th>\n",
       "      <th>lr_means</th>\n",
       "      <th>cellphone_pvals</th>\n",
       "      <th>expr_prod</th>\n",
       "      <th>...</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>weight</th>\n",
       "      <th>comparison</th>\n",
       "      <th>cell_type_y</th>\n",
       "      <th>fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cancer cell</td>\n",
       "      <td>0.149646</td>\n",
       "      <td>2.016727</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>1.754381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.806840</td>\n",
       "      <td>...</td>\n",
       "      <td>69.385540</td>\n",
       "      <td>5.229754</td>\n",
       "      <td>0.562938</td>\n",
       "      <td>9.290112</td>\n",
       "      <td>1.541259e-20</td>\n",
       "      <td>8.716272e-18</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Fibroblast S1</td>\n",
       "      <td>3.883302e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cancer cell</td>\n",
       "      <td>0.149646</td>\n",
       "      <td>2.016727</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>1.754381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.806840</td>\n",
       "      <td>...</td>\n",
       "      <td>50.023148</td>\n",
       "      <td>2.409619</td>\n",
       "      <td>1.000612</td>\n",
       "      <td>2.408145</td>\n",
       "      <td>1.603383e-02</td>\n",
       "      <td>9.992474e-01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>pDC</td>\n",
       "      <td>3.615409e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cancer cell</td>\n",
       "      <td>0.149646</td>\n",
       "      <td>2.016727</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>1.754381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.806840</td>\n",
       "      <td>...</td>\n",
       "      <td>13.649447</td>\n",
       "      <td>2.251672</td>\n",
       "      <td>0.692260</td>\n",
       "      <td>3.252638</td>\n",
       "      <td>1.143390e-03</td>\n",
       "      <td>1.174123e-02</td>\n",
       "      <td>1.065627</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>3.478185e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cancer cell</td>\n",
       "      <td>0.149646</td>\n",
       "      <td>2.016727</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>1.754381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.806840</td>\n",
       "      <td>...</td>\n",
       "      <td>293.180328</td>\n",
       "      <td>0.349132</td>\n",
       "      <td>0.449523</td>\n",
       "      <td>0.776673</td>\n",
       "      <td>4.373518e-01</td>\n",
       "      <td>9.993719e-01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>DC3</td>\n",
       "      <td>5.781782e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cancer cell</td>\n",
       "      <td>0.149646</td>\n",
       "      <td>2.016727</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>1.754381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.806840</td>\n",
       "      <td>...</td>\n",
       "      <td>3.244034</td>\n",
       "      <td>2.225004</td>\n",
       "      <td>0.694730</td>\n",
       "      <td>3.202687</td>\n",
       "      <td>1.361520e-03</td>\n",
       "      <td>1.685540e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>gamma</td>\n",
       "      <td>4.067015e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17732</th>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>0.110253</td>\n",
       "      <td>2.485961</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>GPC4</td>\n",
       "      <td>0.762289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.145144</td>\n",
       "      <td>...</td>\n",
       "      <td>13.649193</td>\n",
       "      <td>1.316913</td>\n",
       "      <td>0.198139</td>\n",
       "      <td>6.646419</td>\n",
       "      <td>3.003087e-11</td>\n",
       "      <td>1.300155e-10</td>\n",
       "      <td>0.753878</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>3.026589e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17733</th>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>0.110253</td>\n",
       "      <td>2.485961</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>GPC4</td>\n",
       "      <td>0.762289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.145144</td>\n",
       "      <td>...</td>\n",
       "      <td>40.516931</td>\n",
       "      <td>0.858156</td>\n",
       "      <td>0.155373</td>\n",
       "      <td>5.523202</td>\n",
       "      <td>3.328775e-08</td>\n",
       "      <td>8.686158e-07</td>\n",
       "      <td>0.884606</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Endothelial lymphatic</td>\n",
       "      <td>2.366902e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17734</th>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>0.110253</td>\n",
       "      <td>2.485961</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>GPC4</td>\n",
       "      <td>0.762289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.145144</td>\n",
       "      <td>...</td>\n",
       "      <td>3.901726</td>\n",
       "      <td>-0.879704</td>\n",
       "      <td>0.364658</td>\n",
       "      <td>-2.412410</td>\n",
       "      <td>1.584745e-02</td>\n",
       "      <td>6.112068e-02</td>\n",
       "      <td>0.798758</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Plasmablast</td>\n",
       "      <td>3.587344e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17735</th>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>0.110253</td>\n",
       "      <td>2.485961</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>GPC4</td>\n",
       "      <td>0.762289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.145144</td>\n",
       "      <td>...</td>\n",
       "      <td>6.870759</td>\n",
       "      <td>-0.607168</td>\n",
       "      <td>0.356138</td>\n",
       "      <td>-1.704870</td>\n",
       "      <td>8.821873e-02</td>\n",
       "      <td>1.802170e-01</td>\n",
       "      <td>1.091076</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Plasma IgG</td>\n",
       "      <td>1.580301e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17736</th>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>0.110253</td>\n",
       "      <td>2.485961</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>Schwann cell</td>\n",
       "      <td>COL18A1</td>\n",
       "      <td>GPC4</td>\n",
       "      <td>0.762289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.145144</td>\n",
       "      <td>...</td>\n",
       "      <td>1.966230</td>\n",
       "      <td>-0.186462</td>\n",
       "      <td>2.184440</td>\n",
       "      <td>-0.085359</td>\n",
       "      <td>9.319758e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.760208</td>\n",
       "      <td>tumor_vs_normal</td>\n",
       "      <td>Enteroendocrine</td>\n",
       "      <td>9.601422e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17737 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cell_type_x  fraction_expressed  expr_mean      source        target  \\\n",
       "0       Cancer cell            0.149646   2.016727  Neutrophil  Schwann cell   \n",
       "1       Cancer cell            0.149646   2.016727  Neutrophil  Schwann cell   \n",
       "2       Cancer cell            0.149646   2.016727  Neutrophil  Schwann cell   \n",
       "3       Cancer cell            0.149646   2.016727  Neutrophil  Schwann cell   \n",
       "4       Cancer cell            0.149646   2.016727  Neutrophil  Schwann cell   \n",
       "...             ...                 ...        ...         ...           ...   \n",
       "17732  Schwann cell            0.110253   2.485961    Pericyte  Schwann cell   \n",
       "17733  Schwann cell            0.110253   2.485961    Pericyte  Schwann cell   \n",
       "17734  Schwann cell            0.110253   2.485961    Pericyte  Schwann cell   \n",
       "17735  Schwann cell            0.110253   2.485961    Pericyte  Schwann cell   \n",
       "17736  Schwann cell            0.110253   2.485961    Pericyte  Schwann cell   \n",
       "\n",
       "      source_genesymbol target_genesymbol  lr_means  cellphone_pvals  \\\n",
       "0                 CXCL8              SDC3  1.754381              0.0   \n",
       "1                 CXCL8              SDC3  1.754381              0.0   \n",
       "2                 CXCL8              SDC3  1.754381              0.0   \n",
       "3                 CXCL8              SDC3  1.754381              0.0   \n",
       "4                 CXCL8              SDC3  1.754381              0.0   \n",
       "...                 ...               ...       ...              ...   \n",
       "17732           COL18A1              GPC4  0.762289              0.0   \n",
       "17733           COL18A1              GPC4  0.762289              0.0   \n",
       "17734           COL18A1              GPC4  0.762289              0.0   \n",
       "17735           COL18A1              GPC4  0.762289              0.0   \n",
       "17736           COL18A1              GPC4  0.762289              0.0   \n",
       "\n",
       "       expr_prod  ...    baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0       0.806840  ...   69.385540        5.229754  0.562938  9.290112   \n",
       "1       0.806840  ...   50.023148        2.409619  1.000612  2.408145   \n",
       "2       0.806840  ...   13.649447        2.251672  0.692260  3.252638   \n",
       "3       0.806840  ...  293.180328        0.349132  0.449523  0.776673   \n",
       "4       0.806840  ...    3.244034        2.225004  0.694730  3.202687   \n",
       "...          ...  ...         ...             ...       ...       ...   \n",
       "17732   0.145144  ...   13.649193        1.316913  0.198139  6.646419   \n",
       "17733   0.145144  ...   40.516931        0.858156  0.155373  5.523202   \n",
       "17734   0.145144  ...    3.901726       -0.879704  0.364658 -2.412410   \n",
       "17735   0.145144  ...    6.870759       -0.607168  0.356138 -1.704870   \n",
       "17736   0.145144  ...    1.966230       -0.186462  2.184440 -0.085359   \n",
       "\n",
       "             pvalue          padj    weight       comparison  \\\n",
       "0      1.541259e-20  8.716272e-18  1.000000  tumor_vs_normal   \n",
       "1      1.603383e-02  9.992474e-01  1.000000  tumor_vs_normal   \n",
       "2      1.143390e-03  1.174123e-02  1.065627  tumor_vs_normal   \n",
       "3      4.373518e-01  9.993719e-01  1.000000  tumor_vs_normal   \n",
       "4      1.361520e-03  1.685540e-02  1.000000  tumor_vs_normal   \n",
       "...             ...           ...       ...              ...   \n",
       "17732  3.003087e-11  1.300155e-10  0.753878  tumor_vs_normal   \n",
       "17733  3.328775e-08  8.686158e-07  0.884606  tumor_vs_normal   \n",
       "17734  1.584745e-02  6.112068e-02  0.798758  tumor_vs_normal   \n",
       "17735  8.821873e-02  1.802170e-01  1.091076  tumor_vs_normal   \n",
       "17736  9.319758e-01  1.000000e+00  0.760208  tumor_vs_normal   \n",
       "\n",
       "                 cell_type_y           fdr  \n",
       "0              Fibroblast S1  3.883302e-19  \n",
       "1                        pDC  3.615409e-02  \n",
       "2               Schwann cell  3.478185e-03  \n",
       "3                        DC3  5.781782e-01  \n",
       "4                      gamma  4.067015e-03  \n",
       "...                      ...           ...  \n",
       "17732             Epithelial  3.026589e-10  \n",
       "17733  Endothelial lymphatic  2.366902e-07  \n",
       "17734            Plasmablast  3.587344e-02  \n",
       "17735             Plasma IgG  1.580301e-01  \n",
       "17736        Enteroendocrine  9.601422e-01  \n",
       "\n",
       "[17737 rows x 28 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpdb_sig_int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "217ee1d9-6621-431b-84ea-064add1e3231",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.histplot(data=cpdb_sig_int, x=\"log2FoldChange\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e5ca234-4f34-4594-b4be-1e8a1ee86ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(cpdb_sig_int.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "366af7e0-7752-418e-bb18-3800b633b502",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This is input for CIRCOS PLOT \n",
    "cpdb_sig_int.to_csv(\"epithelial_cancer_old_atlas.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54d88101-32d8-4d83-924a-0fa924096600",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72f1838-c3d1-47b5-a6c2-4d19465f0da5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "19931ef0-2451-4764-aebe-a47c1a7c1ec3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Che_2021-SRR14710618-AAACCTGAGCGATAGC     tumor\n",
       "Che_2021-SRR14710618-AAACCTGAGCGTGTCC     tumor\n",
       "Che_2021-SRR14710618-AAACCTGCAAGGTGTG     tumor\n",
       "Che_2021-SRR14710618-AAACCTGCACAGCCCA     tumor\n",
       "Che_2021-SRR14710618-AAACCTGCACGACTCG     tumor\n",
       "                                          ...  \n",
       "Zheng_2022-GSM4904245-TTTGGTTGTCTCTCTG    tumor\n",
       "Zheng_2022-GSM4904245-TTTGGTTTCCTAGAAC    tumor\n",
       "Zheng_2022-GSM4904245-TTTGGTTTCGGTTAAC    tumor\n",
       "Zheng_2022-GSM4904245-TTTGTCAAGTGTTGAA    tumor\n",
       "Zheng_2022-GSM4904245-TTTGTCAAGTTAGGTA    tumor\n",
       "Name: sample_type, Length: 541643, dtype: category\n",
       "Categories (1, object): ['tumor']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.sample_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b99335-930d-4b96-ad5f-660d6f04b597",
   "metadata": {},
   "source": [
    "pseudobulk and deseq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d0cf8e9d-3a2f-47b5-89fd-995fd2466b9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/liana_2024/lib/python3.11/site-packages/decoupler/utils_anndata.py:181: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "/home/kvalem/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:430: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n"
     ]
    }
   ],
   "source": [
    "pdata = dc.get_pseudobulk(\n",
    "    adata,\n",
    "    sample_col='sample_id',\n",
    "    groups_col=['sample_type',\"cell_type\"],\n",
    "    layer='counts',\n",
    "    mode='sum',\n",
    "    min_cells=0,\n",
    "    min_counts=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "de73264c-7769-49a3-a1d2-99d5df6e3a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "resDir = '/data/projects/2022/CRCA/results/v1/final/liana_cell2cell/core_atlas/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bbd1ec19-48bc-4bbf-a9a9-d92dcae97716",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison=\"tumor_normal\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "146e6e54-149d-42a5-b571-4828f91c325c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata.var_names.name = \"gene_id\"\n",
    "\n",
    "colData = pdata.obs\n",
    "colData.index.name = \"sample_col\"\n",
    "\n",
    "colData.to_csv(f\"{resDir}/02_pseudobulk/{comparison}_colData.csv\")\n",
    "rowData = pdata.var[[\"Geneid\", \"GeneSymbol\", \"Chromosome\", \"Class\", \"Length\"]]\n",
    "rowData.to_csv(f\"{resDir}/02_pseudobulk/{comparison}_rowData.csv\")\n",
    "count_mat = pdata.to_df().T\n",
    "count_mat.index.name = \"gene_id\"\n",
    "count_mat.to_csv(f\"{resDir}/02_pseudobulk/{comparison}_count_mat.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8eab86-ce74-460e-8da9-25638c1d6112",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "liana_2024",
   "language": "python",
   "name": "liana_2024"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
